src/sql/store.py

```diff
@@ -83,9 +83,12 @@ def __init__(self, store: SQLStore, query: str, with_: Iterable = None):
 
     def __str__(self) -> str:
         with_all = _get_dependencies(self._store, self._with_)
-        return _template.render(
+        rendered_template = _template.render(
             query=self._query, saved=self._store._data, with_=with_all
         )
+        if sql.connection.Connection.current:
+            return sql.connection.Connection.current._transpile_query(rendered_template)
```

Comment:
we have an edge case where. because the SQLStore is combining SQL code from two sources. Given a CTE:

```sql
with something as (
    select * from ...
), another as (
    select * from ...
)
select * from ...
```

we're writing the CTEs (`with something as (...)`) on behalf of the user, but the user is supplying the inner `select * from ...` queries. So I think we need a bit to tweak the logic a bit here. we need to transpile the [template](https://github.com/ploomber/jupysql/blob/ecd91c1a598a2c9f1beeecf5c17474795b64a574/src/sql/store.py#LL66) by using duckdb as the source dialect and the user's dialect as target, and then, render the template with jinja by adding the `select * from` statements unchanged since they're supplied by the user and we should not transpile user-submitted code - let's add a test case on this

Reply:
I have some feedbacks.

### Here is the plan:
1. Use sqlglot to transpile the template, which contains some unparsed raw sql clauses supplied by user, something like `WITH "with_query" AS ({{save_query_with_query}}) {{user_query}}`
2. After it's transpiled, use jinja to populate the raw sql clauses supplied by user

However, I was getting some error on step one, when I executed:
```
query = """
WITH with_query AS ({{save_query_with_query}}) {{user_query}}
"""
print(sqlglot.transpile(query, read="duckdb", write="mysql")[0])
```
Shows error, seems we can't transpile the query clause with {{variable}} 
```
ParseError: struct does not support CTE.
  
WITH with_query AS ({{save_query_with_query}}) {{user_query}}
```

Example of query clause without {{variable}}, this executes fine
```
query = """
WITH with_query AS (SELECT * FROM test) SELECT * from with_query
"""
print(sqlglot.transpile(query, read="duckdb", write="mysql")[0])
```


## Another possible workaround solution
https://sqlglot.com/sqlglot.html#build-and-modify-sql

We might write the sql by using sqlglot expression, then parse to the target dialect, which might be a more generic way to define our internal sql clause (even possible to solve the problem with [PercentileDisc](https://sqlglot.com/sqlglot/expressions.html#PercentileDisc))

I am still doing some research on this

Reply:
### Update - workaround solution
The [latest solution is here](https://github.com/ploomber/jupysql/pull/164/files#diff-a3d1291b54f0d59f191101a81a116bad5859a2bd81603b9b7c1cbea77d18cc10R74-R81), I abandoned the jinja template method since we can natively construct the query with saved `with clause`

So we want to construct the final query with some predefined `with clause` like this:
```
WITH sub_query_a AS (SELECT * FROM table_a) 
WITH sub_query_b AS (SELECT * FROM table_b) 
SELECT * FROM sub_query_a 
```

The constraint is we want to "keep" all user-supplied queries untouched, so translate to the sqlglot fashion will be like this
`parse_one("SELECT * FROM sub_query_a").with_("sub_query_a", "SELECT * FROM table_a").with_("sub_query_b", "SELECT * FROM table_b").sql()`


I think this way is the best way to keep all user-supplied statements, since we already use the native way to construct the user query with `with clause`. If we provide the dialect attribute in `parse_one` or `with_`, it might transpile the user statement though

Passed the integration testing among all databases, but failed on the doc build now

Ref: I asked [the question](https://tobiko-data.slack.com/archives/C0448SFS3PF/p1677792457879019) in sqlglot slack channel

Reply:
nvm. we still need jinja for the new variable substitution format